beacon/resources/beacon.py

```diff
@@ -33,73 +34,73 @@ def main(request, response):
+    if b"origin" in request.GET:
+        response.headers.set(b"Access-Control-Allow-Origin",
+                             request.GET.first(b"origin"))
+    if b"credentials" in request.GET:
+        response.headers.set(b"Access-Control-Allow-Credentials",
+                             request.GET.first(b"credentials"))
 
     # Handle the 'store' and 'stat' commands.
-    if command == "store":
+    if command == b"store":
         error = None
 
         # Only store the actual POST requests, not any preflight/OPTIONS
         # requests we may get.
-        if request.method == "POST":
-            payload = ""
-            if "Content-Type" in request.headers and \
-               "form-data" in request.headers["Content-Type"]:
-                if "payload" in request.POST:
+        if request.method == u"POST":
+            payload = b""
+            if b"Content-Type" in request.headers and \
+               b"form-data" in request.headers[b"Content-Type"]:
+                if b"payload" in request.POST:
                     # The payload was sent as a FormData.
-                    payload = request.POST.first("payload")
+                    payload = request.POST.first(b"payload")
                 else:
                     # A FormData was sent with an empty payload.
                     pass
             else:
                 # The payload was sent as either a string, Blob, or BufferSource.
                 payload = request.body
 
-            payload_parts = filter(None, payload.split(":"))
+            payload_parts = list(filter(None, isomorphic_decode(payload).split(u":")))
             if len(payload_parts) > 0:
                 payload_size = int(payload_parts[0])
 
                 # Confirm the payload size sent matches with the number of
                 # characters sent.
                 if payload_size != len(payload_parts[1]):
-                    error = "expected %d characters but got %d" % (
+                    error = u"expected %d characters but got %d" % (
                         payload_size, len(payload_parts[1]))
                 else:
                     # Confirm the payload contains the correct characters.
                     for i in range(0, payload_size):
-                        if payload_parts[1][i] != "*":
-                            error = "expected '*' at index %d but got '%s''" % (
+                        if payload_parts[1][i] != u"*":
+                            error = u"expected '*' at index %d but got '%s''" % (
                                 i, payload_parts[1][i])
                             break
 
             # Store the result in the stash so that it can be retrieved
             # later with a 'stat' command.
-            request.server.stash.put(id, {"error": error})
-        elif request.method == "OPTIONS":
+            request.server.stash.put(id, {u"error": error})
+        elif request.method == u"OPTIONS":
```

Comment:
Why is method not a byte sequence?

Reply:
`request.method` is actually unfortunately still a native string. This just happens to work because in Python 2 unicode can be compared implicitly with str.

```suggestion
        elif isomorphic_encode(request.method) == b"OPTIONS":
```
is probably better

Reply:
I don't think that's better. Ideally we change request.method, but if we can't than we should use `u` here. Probably a follow-up issue.

Reply:
Updated. Thank you!

Reply:
This is https://github.com/web-platform-tests/wpt/issues/23362. I'm marking the thread as resolved for now.